### PR TITLE
chore: allow parenthesizing in two type locations 

### LINKED ISSUE #2375: (fuzzer) - The parser does not allow parenthesizing in two type locations
### Aim

I tried to compile types with parentheses in two locations:
* `str<(Foo)>` - string type with a parethesized argument
* `[Foo; (Bar)]` - array type with a parenthesized length

### Expected Behavior

Expected the types to compile with at most a warning.

### Bug

This is inconsistent with other type expressions in Noir, and with the Rust parser which gives a warning (I know we differ from Rust grammar, but still).

### To Reproduce

Compile:
```rust
fn foo<Bar>(bar: [Bar; (123)]) {}
```
-> Error: `Expression is invalid in an array-length type: 'Error'. Only unsigned integer constants, globals, generics, +, -, *, /, and % may be used in this context.`
```rust
fn foo<Foo>(bar: str<(Foo)>) {}
```
-> ???

### Installation Method

Compiled from source

### Nargo Version

_No response_

### Additional Context

Bug found with the parser-fuzzer tool.

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by f01dab1e**
# Description

allow parenthesizing in two type locations:

```
fn foo2<Bar>(bar: [Bar; (123)]) {}

fn foo<Foo>(bar: str<(Foo)>) {}
```

## Problem

Resolves #2375 

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_frontend/src/parser/parser.rs [hunk 1]
```diff
@@ -942,10 +942,11 @@ fn type_expression() -> impl NoirParser<UnresolvedTypeExpression> {
         expression_with_precedence(
             Precedence::lowest_type_precedence(),
             nothing(),
-            expr,
+            expr.clone(),
             true,
             false,
         )
+        .or(parenthesized(expr))
```
Comment (jfecher) [inline crates/noirc_frontend/src/parser/parser.rs:949]: This shouldn't be needed, it should be handled within `type_expression_atom`

File: crates/noirc_frontend/src/parser/parser.rs [hunk 2]
```diff
@@ -1056,13 +1057,13 @@ where
...
+            type_expression_term(expr_parser).boxed()
         } else {
-            term(expr_parser, expr_no_constructors, allow_constructors)
-                .boxed()
-                .labelled(ParsingRuleLabel::Term)
-        }
+            term(expr_parser, expr_no_constructors, allow_constructors).boxed()
+        };
+
+        expr_parser.clone().or(parenthesized(expr_parser)).boxed().labelled(ParsingRuleLabel::Term)
```
Comment (jfecher) [inline crates/noirc_frontend/src/parser/parser.rs:1066]: The `type_expression_atom` parser should already support parenthesized type expressions, do you know why we need this extra check?

**Review by jfecher — CHANGES_REQUESTED** [review]

File: crates/noirc_frontend/src/parser/parser.rs [hunk 3]
```diff
@@ -941,8 +941,8 @@ fn type_expression() -> impl NoirParser<UnresolvedTypeExpression> {
     recursive(|expr| {
         expression_with_precedence(
             Precedence::lowest_type_precedence(),
-            nothing(),
             expr,
+            nothing(),
```
Comment (jfecher) [inline crates/noirc_frontend/src/parser/parser.rs:945]: Can't believe this wasn't caught earlier, thanks for finding the bug!

**Review by jfecher — APPROVED** [review]

File: crates/noirc_frontend/src/parser/parser.rs (hunk 3 above)
Comment (pczarn) [reply crates/noirc_frontend/src/parser/parser.rs:945]: @jfecher :smiley: :shipit:

**Pull request merged**